pkg/datapath/loader/tcx.go

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package loader
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+
+	"github.com/cilium/cilium/pkg/bpf"
+)
+
+func parentToAttachType(parent uint32) ebpf.AttachType {
+	switch parent {
+	case netlink.HANDLE_MIN_INGRESS:
+		return ebpf.AttachTCXIngress
+	case netlink.HANDLE_MIN_EGRESS:
+		return ebpf.AttachTCXEgress
+	}
+	panic(fmt.Sprintf("invalid tc direction: %d", parent))
+}
+
+// upsertTCXProgram updates or creates a new tcx attachment for prog to device.
+// Returns [link.ErrNotSupported] if tcx is not supported on the node.
+func upsertTCXProgram(device netlink.Link, prog *ebpf.Program, progName, bpffsDir string, parent uint32) error {
+	err := updateTCX(prog, progName, bpffsDir)
+	if err == nil {
+		// Link was updated, nothing left to do.
+		return nil
+	}
+	if !errors.Is(err, os.ErrNotExist) {
+		// Unrecoverable error, surface to the caller.
+		return fmt.Errorf("updating tcx program: %w", err)
+	}
+
+	return attachTCX(device, prog, progName, bpffsDir, parentToAttachType(parent))
+}
+
+// attachTCX creates a new tcx attachment for prog to device at the given attach
+// type. It pins the resulting link object to progName in bpffsDir.
+//
+// progName is typically the Program's key in CollectionSpec.Programs.
+func attachTCX(device netlink.Link, prog *ebpf.Program, progName, bpffsDir string, attach ebpf.AttachType) error {
+	l, err := link.AttachTCX(link.TCXOptions{
+		Program:   prog,
+		Attach:    attach,
+		Interface: device.Attrs().Index,
+		Anchor:    link.Tail(),
+	})
+	if err != nil {
+		return fmt.Errorf("attaching tcx: %w", err)
+	}
+
+	defer func() {
+		// The program was successfully attached using tcx. Closing a link does not
+		// detach the program if the link is pinned.
+		if err := l.Close(); err != nil {
+			log.Warnf("Failed to close tcx link for program %s", progName)
+		}
+	}()
+
+	pin := filepath.Join(bpffsDir, progName)
+	if err := l.Pin(pin); err != nil {
+		return fmt.Errorf("pinning link at %s for program %s : %w", pin, progName, err)
+	}
+
+	log.Infof("Program %s attached to device %s using tcx", progName, device.Attrs().Name)
+
+	return nil
+}
+
+// updateTCX attempts to update an existing tcx link called progName in
+// bpffsDir. If the link is defunct, the pin is removed.
+//
+// Returns nil if the update was successful. Returns an error wrapping
+// [os.ErrNotExist] if the link is defunct or missing.
+func updateTCX(prog *ebpf.Program, progName, bpffsDir string) error {
+	// Attempt to open and update an existing link.
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UpdateLink(pin, prog)
+	switch {
+	// Update successful, nothing left to do.
```

Comment:
this is all a bit odd to me. I would either return from all case statements or none. But just a nitpick.

Reply:
Yeah it does look weird, I'll fix.